src/output.js

```diff
@@ -3,36 +3,90 @@ const path = require("path");
 const { config } = require("./config");
 
 function saveToCSV(filename, data) {
-  const headers = Object.keys(data[0]);
-  const filteredHeaders = headers.filter((header) => !header.startsWith("_"));
-  const headerRow = filteredHeaders.join(",") + "\n";
-  const dataRows = data
-    .map((row) => filteredHeaders.map((header) => row[header]).join(","))
-    .join("\n");
-  const csv = headerRow + dataRows;
-  writeFile(filename, csv);
+  return new Promise((resolve, reject) => {
+    if (!data || data.length === 0) {
+      console.error(`No data provided to write to ${filename}`);
+      return reject(new Error("Data is empty."));
+    }
+
+    const headers = Object.keys(data[0]);
+    const filteredHeaders = headers.filter((header) => !header.startsWith("_"));
+
+    // Create a write stream
+    const stream = fs.createWriteStream(filename);
+
+    // Handle stream errors and log them
+    stream.on('error', (err) => {
+      console.error(`Error writing to CSV file ${filename}: ${err.message}`);
+      reject(err);
+    });
+
+    // Write the headers first
+    stream.write(filteredHeaders.join(",") + "\n");
+
+    // Write data rows incrementally
+    data.forEach((row) => {
+      const rowData = filteredHeaders.map((header) => row[header] || "").join(",");
+      stream.write(rowData + "\n");
+    });
+
+    stream.end(() => {
+      resolve();
+    });
+  });
 }
 
 function writeFile(filename, data, opt) {
-  const dir = path.dirname(filename);
-  if (!fs.existsSync(dir)) {
-    fs.mkdirSync(dir, { recursive: true });
-  }
-  fs.writeFileSync(filename, data, opt);
-}
+  return new Promise((resolve, reject) => {
+    const dir = path.dirname(filename);
+    if (!fs.existsSync(dir)) {
+      fs.mkdirSync(dir, { recursive: true });
+    }
 
-function deleteFile(filename) {
-  fs.access(filename, fs.constants.F_OK, (err) => {
-    if (err) {
-        return; //File doesn't exist. Ignore silently
+    // Create a write stream
+    const stream = fs.createWriteStream(filename, opt);
+
+    // Handle stream errors
+    stream.on('error', (err) => {
+      console.error(`Error writing to file ${filename}: ${err.message}`);
+      reject(err);
+    });
+
+    // Write to file
+    if (Array.isArray(data)) {
+      data.forEach((line) => {
+        stream.write(line + '\n');
+      });
+    }
+    else {
+      stream.write(data);
     }
 
-    fs.unlink(filename, (err) => {
-        if (err) {
-            console.error(`Error deleting file: ${err}`);
-        }
+    // End the stream
+    stream.end(() => {
+      resolve();
     });
-});
+  });
+}
+
+function deleteFile(filename) {
+  try {
+    // Check if file exists
+    fs.accessSync(filename, fs.constants.F_OK);
+
+    // If file exists, proceed with unlinking it
+    fs.unlinkSync(filename);
+  }
+  catch (err) {
+    // Handle error if file does not exist or other issues
```

Comment:
This comment isn't necessary because it's obvious from the line below that the error we caught was for "file not existing," which we're ignoring.